3rd-party/class.jetpack-amp-support.php

```diff
@@ -361,28 +364,84 @@ public static function render_sharing_html( $markup, $sharing_enabled ) {
 		if ( empty( $sharing_enabled ) ) {
 			return $markup;
 		}
-		$supported_services = array(
-			'facebook'  => array(
-				/** This filter is documented in modules/sharedaddy/sharing-sources.php */
-				'data-param-app_id' => apply_filters( 'jetpack_sharing_facebook_app_id', '249643311490' ),
-			),
-			'twitter'   => array(),
-			'pinterest' => array(),
-			'whatsapp'  => array(),
-			'tumblr'    => array(),
-			'linkedin'  => array(),
-		);
-		$sharing_links      = array();
+
+		$sharing_links = array();
 		foreach ( $sharing_enabled['visible'] as $id => $service ) {
-			if ( ! isset( $supported_services[ $id ] ) ) {
-				$sharing_links[] = "<!-- not supported: $id -->";
+			$args = array();
+			$skip = false;
+			switch ( $service->get_id() ) {
```

Comment:
Part of me wonders if at this point, we could include this in each service's class (e.g. `Share_Facebook` [#](https://github.com/Automattic/jetpack/blob/3699f34bb90a2b0a2af221042dacacf777238d2b/modules/sharedaddy/sharing-sources.php#L955)) as a new method, to avoid duplicating things and make it clear what services are supported by seeing which ones add the method to their class. What do you think of such an approach?

Reply:
I actually had a similar thought pattern, but then when I looked at the code further, I noticed that all AMP functionality is abstracted away in to this single file/class. I figured that was a design decision and decided to leave anything that deals with AMP in to this file.

I originally had the AMP CSS being enqueued inside the Sharedaddy class too, but moved it all to this file.

I'm happy to move AMP related functionality in to the modules, but it would break what I believe to be an intentional design of having it abstracted to a single place. Would love to hear your thoughts with regards to this?

Reply:
Yes, it was a conscious decision when we first worked on AMP compatibility. I think it's worth revisiting now that we will support more sharing services; at this point AMP support for the sharing module has become significant enough that we could include it in the sharing file, as long as we are clear on what the methods are used for in there.

What do you think?

Reply:
Makes sense. I'd propose that we include a method which renders the AMP markup like `get_display` does in the parent class. This can return false if AMP is not supported for a given module or if it needs customizations (as per the rules here), we can override the method in the individual service. I'll work on this and send a commit to the PR to get your thoughts.

Reply:
@jeherve Implemented in: https://github.com/Automattic/jetpack/pull/15497/commits/d27daaecc836373885a4e9825865499c8de1d616 Would appreciate your thoughts